File: src/main/java/info/phj233/onlinechat/controller/UserController.java
package info.phj233.onlinechat.controller;

import info.phj233.onlinechat.config.JWTConfig;
import info.phj233.onlinechat.model.User;
import info.phj233.onlinechat.model.UserDetailImpl;
import info.phj233.onlinechat.model.value.Register;
import info.phj233.onlinechat.service.UserService;
import info.phj233.onlinechat.util.JWTUtil;
import info.phj233.onlinechat.util.result.E;
import info.phj233.onlinechat.util.result.R;
import jakarta.servlet.http.HttpServletRequest;
import lombok.RequiredArgsConstructor;
import org.springframework.data.domain.Page;
import org.springframework.security.access.prepost.PreAuthorize;
import org.springframework.security.core.Authentication;
import org.springframework.web.bind.annotation.*;
import org.springframework.web.multipart.MultipartFile;

import java.util.HashMap;
import java.util.Map;

/**
 * @author phj233
 * @since  2023/3/10 8:59
 * @version 1.0
 */
@RestController
@CrossOrigin
@RequiredArgsConstructor
@RequestMapping("/user")
public class UserController {
    private final UserService userService;
    /**
     * 注册
     * @param register 注册信息
     * @return Result
     * @see Register
     */
    @PostMapping("/register")
    public R<User> register(@RequestBody Register register){
        return userService.register(register);
    }

    /**
     * 获取用户信息
     * @param authentication 用户信息
     * @return Result
     */
    @GetMapping("/info")
    public R<Object> info(Authentication authentication) {
        if (authentication == null) {
            return R.error(E.ACCESS_DENIED, "未认证");
        }
        UserDetailImpl user = (UserDetailImpl) authentication.getPrincipal();
        Map<String, Object> map = new HashMap<>();
        map.put("userinfo", user.getUser());
        map.put("roles", user.getAuthorities());
        return R.ok(map);
    }

    /**
     * 验证token是否过期
     * @param request 请求
     * @return Result
     */
    @GetMapping("/checktoken")
    public R<Boolean> checktoken(HttpServletRequest request) {
        String token = request.getHeader(JWTConfig.tokenHeader);
        return R.ok(JWTUtil.verifyToken(token));
    }

    /**
     * 修改用户信息
     * @param user 用户信息
     * @return Result
     */
    @PreAuthorize("hasAnyAuthority('admin', 'USER')")
    @PostMapping("/update")
    public R<User> update(@RequestBody User user) {
        return userService.update(user);
    }

    /**
     * 修改用户头像
     * @param file 头像文件
     * @param authentication 用户信息
     * @param request 请求
     * @return Result
     */
    @PostMapping("/avatar")
    public R<Object> avatar(@RequestParam("file")MultipartFile file, Authentication authentication, HttpServletRequest request) {
        return userService.updateAvatar(file, authentication, request);
    }

    /**
     * 分页获取所有用户
     * @return Result
     */
    @GetMapping("/page")
    public R<Page<User>> pagefindAll(@RequestParam(value = "page", defaultValue = "0") Integer page,
                               @RequestParam(value = "size", defaultValue = "10") Integer size) {
        return userService.pagefindAll(page, size);
    }



}

File: src/main/kotlin/top/phj233/backend_sb3/controller/UserController.kt
package top.phj233.backend_sb3.controller

import cn.dev33.satoken.annotation.SaCheckRole
import cn.dev33.satoken.annotation.SaIgnore
import cn.dev33.satoken.secure.BCrypt
import cn.dev33.satoken.stp.StpUtil
import jakarta.servlet.http.HttpServletRequest
import org.babyfish.jimmer.client.EnableImplicitApi
import org.babyfish.jimmer.sql.ast.mutation.SaveMode
import org.slf4j.Logger
import org.slf4j.LoggerFactory
import org.springframework.data.domain.Page
import org.springframework.data.domain.PageRequest
import org.springframework.http.ResponseEntity
import org.springframework.web.bind.annotation.*
import top.phj233.backend_sb3.exception.MessageException
import top.phj233.backend_sb3.model.dto.UserLogin
import top.phj233.backend_sb3.model.dto.UserRegister
import top.phj233.backend_sb3.model.dto.UserUpdate
import top.phj233.backend_sb3.model.dto.UserView
import top.phj233.backend_sb3.repository.UserRepository
import top.phj233.backend_sb3.util.ApiResponse
import top.phj233.backend_sb3.util.ResponseUtil

@EnableImplicitApi
@RestController
@RequestMapping("/user")
class UserController(
    val userRepository: UserRepository,
) {
    val log:Logger = LoggerFactory.getLogger(this::class.java)

    /**
     * 登录
     * @param userLogin 用户登录信息
     * @return 登录结果
     */
    @PostMapping("/login")
    @SaIgnore
    fun login(@RequestBody userLogin : UserLogin): ResponseEntity<ApiResponse<UserView>> {
        log.info("login: $userLogin")
        val userEntity = userRepository.findUserByEmail(userLogin.email)
        if (userEntity != null) {
            if (BCrypt.checkpw(userLogin.password,userEntity.password)) {
                userEntity.email.compareTo(userLogin.email).let {
                    if (it != 0) {
                        throw RuntimeException("邮箱错误")
                    }
                }
                StpUtil.login(userEntity.id)
                val userView = userRepository.findByEmail(userLogin.email, UserView::class)
                return ResponseUtil.success(userView)
            }
        }
        throw RuntimeException("用户名或密码错误")
    }

    @PostMapping("/register")
    @SaIgnore
    fun register(@RequestBody user : UserRegister): ResponseEntity<ApiResponse<String>> {
        val userEntity = userRepository.findUserByUsername(user.username)
        userEntity?.let {
            throw RuntimeException("用户已存在")
        }
        user.copy(password = BCrypt.hashpw(user.password)).let {
            userRepository.save(it,SaveMode.INSERT_ONLY)
            return ResponseUtil.success("注册成功")
        }
    }

    @GetMapping("/logout")
    fun logout(): ResponseEntity<ApiResponse<String>> {
        StpUtil.logout()
        return ResponseUtil.success("登出成功")
    }

    @GetMapping("/info")
    fun info(request: HttpServletRequest): ResponseEntity<ApiResponse<UserView>> {
        log.info(StpUtil.getTokenInfo().toString())
        val userView = userRepository.findByEmail(StpUtil.getLoginIdAsString(), UserView::class)
        return ResponseUtil.success(userView)
    }

    @GetMapping("/list")
    fun list(@RequestParam(defaultValue = "0") page: Int): ResponseEntity<ApiResponse<Page<UserView>>> {
        val users = userRepository.viewer(UserView::class).findAll(PageRequest.of(
            page, 10
        ))
        return ResponseUtil.success(users)
    }

    /**
     * 模糊搜索用户
     * @param username 用户名
     * @return List<UserView> 用户列表
     */
    @GetMapping("/like_search")
    fun likeSearch(@RequestParam username: String,@RequestParam(defaultValue = "0") page: Int): ResponseEntity<ApiResponse<Page<UserView>>> {
        val users = userRepository.findByUsernameLike(username, PageRequest.of(
            page, 10
        ), UserView::class)
        return ResponseUtil.success(users)
    }

    /**
     * 更新用户信息
     * @param userUpdate 用户信息
     * @return UserView 用户信息
     */
    @PutMapping("/update")
    fun update(@RequestBody userUpdate: UserUpdate): ResponseEntity<ApiResponse<UserView>> {
        if (StpUtil.hasRole("admin") || userUpdate.id == StpUtil.getLoginIdAsLong()){
            userUpdate.copy(password = BCrypt.hashpw(userUpdate.password)).let {
                userRepository.update(it)
            }
            return ResponseUtil.success(userUpdate.id?.let { userRepository.viewer(UserView::class).findNullable(it) })
        }
        throw MessageException("权限不足")
    }

    /**
     * 删除用户
     * @param id 用户id
     */
    @SaCheckRole("admin")
    @DeleteMapping("/delete/{id}")
    fun delete(@PathVariable id: Long): ResponseEntity<ApiResponse<String>> {
        val user = userRepository.findById(id).orElse(null)
        user?.let {
            userRepository.delete(it)
            return ResponseUtil.success("删除成功")
        }
        throw MessageException("用户不存在")
    }

}
